api/src/services/roles.ts

```diff
@@ -27,41 +27,116 @@ export class RolesService extends ItemsService {
+						usersUpdated.push(user as Partial<User> & Pick<User, 'id'>);
+					} else {
+						usersAdded.push(user as Partial<User> & Pick<User, 'id'>);
+					}
+				} else {
+					usersCreated.push(user);
+				}
+			}
+
+			usersRemoved.push(...usersBefore.filter((user) => !usersKept.includes(user)));
 		} else {
-			userKeys = users.update.map((user) => user['id']).filter((id) => id);
-		}
+			for (const user of users.update) {
+				if (usersBefore.includes(user['id'])) {
+					usersUpdated.push(user);
+				} else {
+					usersAdded.push(user);
+				}
+			}
 
-		const usersThatWereInRoleBefore = (await this.knex.select('id').from('directus_users').where('role', '=', key)).map(
-			(user) => user.id,
-		);
+			usersCreated.push(...users.create);
+			usersRemoved.push(...users.delete);
+		}
 
-		const usersThatAreRemoved = usersThatWereInRoleBefore.filter((id) =>
-			Array.isArray(users) ? userKeys.includes(id) === false : users.delete.includes(id) === true,
-		);
+		if (role.admin_access === false || role.admin_access === 0) {
+			// Admin users might have moved in from other role, thus becoming non-admin
+			if (usersAdded.length > 0) {
+				const otherAdminUsers = await this.knex
+					.count('*', { as: 'count' })
+					.from('directus_users')
+					.leftJoin('directus_roles', 'directus_users.role', 'directus_roles.id')
+					.whereNotIn('directus_users.id', usersAdded)
+					.andWhere({ 'directus_roles.admin_access': true, status: 'active' })
+					.first();
+
+				const otherAdminUsersCount = +(otherAdminUsers?.count || 0);
+
+				if (otherAdminUsersCount === 0) {
+					throw new UnprocessableContentError({ reason: `You can't remove the last admin user from the admin role` });
+				}
+			}
 
-		const usersThatAreAdded = Array.isArray(users) ? users : users.create;
+			return;
+		}
 
-		// If the role the users are moved to is an admin-role, and there's at least 1 (new) admin
-		// user, we don't have to check for other admin
-		// users
-		if ((role.admin_access === true || role.admin_access === 1) && usersThatAreAdded.length > 0) return;
+		// Only added or created new users
+		if (usersUpdated.length === 0 && usersRemoved.length === 0) return;
+
+		// Active admin user(s) about to be created
+		if (usersCreated.some((user) => !('status' in user) || user.status === 'active')) return;
+
+		const usersDeactivated = [...usersAdded, ...usersUpdated]
+			.filter((user) => 'status' in user && user.status !== 'active')
+			.map((user) => user.id);
+
+		const usersAddedNonDeactivated = usersAdded
+			.filter((user) => !usersDeactivated.includes(user.id))
+			.map((user) => user.id);
+
+		// Active user(s) about to become admin
+		if (
+			usersAddedNonDeactivated.length > 0 &&
+			+(
+				(
+					await this.knex
+						.count('*', { as: 'count' })
+						.from('directus_users')
+						.whereIn('id', usersAddedNonDeactivated)
+						.andWhere({ status: 'active' })
+						.first()
+				)?.count || 0
+			) > 0
+		)
+			return;
```

Comment:
Doing a database call, type casting and fallbacks directly within the if expression is not very readable and prone to errors for future maintenance. The typecasting to Number will fail when postgres returns a BigInt string for huge tables.
```suggestion
		if (usersAddedNonDeactivated.length > 0) {
			const userCount = await this.knex
				.count('*', { as: 'count' })
				.from('directus_users')
				.whereIn('id', usersAddedNonDeactivated)
				.andWhere({ status: 'active' })
				.first()?.count ?? 0;

			if (BigInt(userCount) > 0) {
				return;
			}
```